File: test/unit/console.test.ts
import { describe, expect, it, vi } from 'vitest';
import { LogStore } from '../../src/core/store.ts';
import type { LogEntry, LogPart, ValueNode } from '../../src/core/types.ts';
import {
	applyFormat,
	formatArguments,
	isSafeColor,
	parseConsoleCss
} from '../../src/sources/console/format.ts';
import { createConsole, hookConsole } from '../../src/sources/console/hook.ts';
import { snapshotValue } from '../../src/sources/console/snapshot.ts';

const capture = (value: unknown): ValueNode => snapshotValue(value);

const textOf = (parts: readonly LogPart[]): string => {
	return parts.map((part) => (part.type === 'text' ? part.text : `<${part.value.kind}>`)).join('');
};

const lastEntry = (store: LogStore): LogEntry => {
	return store.at(store.size - 1) as LogEntry;
};

describe('snapshotValue', () => {
	it('captures primitives the way they print', () => {
		expect(snapshotValue(-0)).toEqual({ kind: 'number', value: '-0' });
		expect(snapshotValue(10n)).toEqual({ kind: 'bigint', value: '10' });
		expect(snapshotValue(Symbol('token'))).toEqual({ kind: 'symbol', value: 'Symbol(token)' });
		expect(snapshotValue(undefined)).toEqual({ kind: 'undefined' });
	});

	it('takes the value at call time', () => {
		const user = { name: 'before' };
		const node = snapshotValue(user);

		user.name = 'after';

		expect(node.children?.[0].value.value).toBe('before');
	});

	it('never runs getters', () => {
		const getter = vi.fn(() => 'secret');
		const value = Object.defineProperty({}, 'lazy', { get: getter, enumerable: true });
		const node = snapshotValue(value);

		expect(getter).not.toHaveBeenCalled();
		expect(node.children?.[0].value).toEqual({ kind: 'accessor', accessor: 'get' });
	});

	it('marks circular references', () => {
		const value: Record<string, unknown> = {};

		value.self = value;

		expect(snapshotValue(value).children?.[0].value.kind).toBe('circular');
	});

	it('respects the depth, property, string and node limits', () => {
		const deep = { a: { b: { c: { d: 1 } } } };
		const wide = Object.fromEntries(
			Array.from({ length: 10 }, (_, index) => [`key${index}`, index])
		);
		const depthNode = snapshotValue(deep, { maxDepth: 2 });
		const inner = depthNode.children?.[0].value.children?.[0].value;

		expect(inner?.kind).toBe('object');
		expect(inner?.children).toBeUndefined();
		expect(snapshotValue(wide, { maxProperties: 3 })).toMatchObject({ omitted: 7 });
		expect(snapshotValue('x'.repeat(20), { maxStringLength: 5 })).toMatchObject({
			value: 'xxxxx',
			truncated: 15
		});
		expect(
			snapshotValue(
				Array.from({ length: 50 }, () => ({ x: 1 })),
				{ maxNodes: 10 }
			).children?.length
		).toBeLessThan(10);
	});

	it('recognizes built-in types', () => {
		expect(snapshotValue(new Date(0))).toEqual({ kind: 'date', value: '1970-01-01T00:00:00.000Z' });
		expect(snapshotValue(/a+b/gi)).toEqual({ kind: 'regexp', value: '/a+b/gi' });
		expect(snapshotValue(new Map([['a', 1]]))).toMatchObject({
			kind: 'map',
			size: 1,
			className: 'Map'
		});
		expect(snapshotValue(new Set([1, 2]))).toMatchObject({ kind: 'set', size: 2 });
		expect(snapshotValue(new Uint8Array([1, 2, 3]))).toMatchObject({
			kind: 'array',
			className: 'Uint8Array',
			size: 3
		});
		expect(snapshotValue(new WeakMap())).toEqual({ kind: 'weak', className: 'WeakMap' });
		expect(snapshotValue(class Example {})).toEqual({ kind: 'class', value: 'Example' });
		expect(snapshotValue(function named() {})).toEqual({ kind: 'function', value: 'named' });
	});

	it('keeps the class name of an instance', () => {
		class User {
			name = 'lognal';
		}

		expect(snapshotValue(new User())).toMatchObject({ kind: 'object', className: 'User' });
	});

	it('captures errors with their cause and without the stack header', () => {
		const error = new TypeError('broken', { cause: new Error('root') });
		const node = snapshotValue(error);

		expect(node).toMatchObject({ kind: 'error', className: 'TypeError', value: 'broken' });
		expect(node.stack?.startsWith('TypeError')).toBe(false);
		expect(node.children?.find((child) => child.key === 'cause')?.value).toMatchObject({
			kind: 'error',
			value: 'root'
		});
	});

	it('survives a proxy whose traps throw', () => {
		const hostile = new Proxy(
			{},
			{
				ownKeys: () => {
					throw new Error('no keys');
				}
			}
		);

		expect(() => snapshotValue(hostile)).not.toThrow();
	});
});

describe('format specifiers', () => {
	it('applies %s, %d, %i and %f with the Console Standard conversions', () => {
		const { parts, rest } = applyFormat(
			['%s|%d|%i|%f|%%', 'text', '42.9', -1.5, '3.25', 'extra'],
			capture
		);

		expect(textOf(parts)).toBe('text|42|-1|3.25|%');
		expect(rest).toEqual(['extra']);
	});

	it('inserts values for %o and %O and leaves specifiers without arguments as written', () => {
		const { parts } = applyFormat(['a %o b %O c %s', { x: 1 }, [1]], capture);

		expect(textOf(parts)).toBe('a <object> b <array> c %s');
	});

	it('styles the text after %c with the allowed properties only', () => {
		const { parts } = applyFormat(
			[
				'%cred%c plain',
				'color: red; background: url(https://example.com/x.png) #000; font-weight: bold',
				''
			],
			capture
		);

		expect(parts[0]).toEqual({
			type: 'text',
			text: 'red',
			style: { color: 'red', background: '#000', bold: true }
		});
		expect(parts[1]).toEqual({ type: 'text', text: ' plain' });
	});

	it('rejects colors that could load a resource', () => {
		expect(isSafeColor('rgb(1 2 3 / 50%)')).toBe(true);
		expect(isSafeColor('url(x)')).toBe(false);
		expect(parseConsoleCss('color: url(x)')).toBeUndefined();
	});

	it('joins the remaining arguments with spaces', () => {
		expect(textOf(formatArguments(['count', 3, 'items'], capture))).toBe('count <number> items');
		expect(textOf(formatArguments(['%d%%', 50, 'done'], capture))).toBe('50% done');
		expect(textOf(formatArguments(['only %s'], capture))).toBe('only %s');
	});
});

describe('console recorder', () => {
	it('records levels and group membership', () => {
		const store = new LogStore();
		const log = createConsole(store);

		log.group('outer');
		log.warn('inside');
		log.groupEnd();
		log.error('outside');

		const [group, inside, outside] = store.toArray();

		expect(group).toMatchObject({ kind: 'group', collapsed: false });
		expect(inside).toMatchObject({ level: 'warn', groups: [group.id] });
		expect(outside).toMatchObject({ level: 'error', groups: [] });
	});

	it('counts, resets counts and warns about a missing counter', () => {
		const store = new LogStore({ mergeRepeats: false });
		const log = createConsole(store);

		log.count('clicks');
		log.count('clicks');
		log.countReset('clicks');
		log.count('clicks');
		log.countReset('other');

		expect(store.toArray().map((entry) => textOf(entry.parts))).toEqual([
			'clicks: 1',
			'clicks: 2',
			'clicks: 1',
			"Count for 'other' does not exist"
		]);
	});

	it('measures timers at call time', () => {
		const store = new LogStore();
		const log = createConsole(store);
		const now = vi.spyOn(performance, 'now');

		now.mockReturnValue(1000);
		log.time('load');
		now.mockReturnValue(1250.5);
		log.timeLog('load', 'halfway');
		log.timeEnd('load');
		log.timeEnd('load');
		now.mockRestore();

		expect(store.toArray().map((entry) => textOf(entry.parts))).toEqual([
			'load: 250.5 ms halfway',
			'load: 250.5 ms',
			"Timer 'load' does not exist"
		]);
	});

	it('writes assertion failures the way the standard describes', () => {
		const store = new LogStore({ mergeRepeats: false });
		const log = createConsole(store);

		log.assert(true, 'not shown');
		log.assert(false);
		log.assert(false, 'value is %d', 3);
		log.assert(false, { code: 1 });

		expect(store.toArray().map((entry) => textOf(entry.parts))).toEqual([
			'Assertion failed',
			'Assertion failed: value is 3',
			'Assertion failed <object>'
		]);
	});

	it('draws console.table as a text table that lines up wide characters', () => {
		const store = new LogStore();

		createConsole(store).table([{ name: '김철수', age: 30 }, { name: 'Ann' }]);

		const lines = textOf(lastEntry(store).parts).split('\n');
		const widths = new Set(
			lines.map((line) =>
				[...line].reduce((sum, character) => sum + (/[가-힣]/.test(character) ? 2 : 1), 0)
			)
		);

		expect(lines[1]).toContain('(index)');
		expect(widths.size).toBe(1);
	});

	it('clears the store and notes it', () => {
		const store = new LogStore();
		const log = createConsole(store);

		log.log('before');
		log.clear();

		expect(store.size).toBe(1);
		expect(lastEntry(store)).toMatchObject({ kind: 'system' });
	});

	it('adds the stack of console.trace', () => {
		const store = new LogStore();

		createConsole(store).trace('here');

		const text = textOf(lastEntry(store).parts);

		expect(text.startsWith('here\n')).toBe(true);
		expect(text).toContain('console.test.ts');
	});
});

describe('hookConsole', () => {
	const fakeConsole = () => {
		const calls: unknown[][] = [];
		const target = {
			log: (...args: unknown[]) => calls.push(args),
			warn: (...args: unknown[]) => calls.push(args)
		};

		return { target: target as unknown as Console, calls };
	};

	it('records calls, still runs the original and restores it', () => {
		const { target, calls } = fakeConsole();
		const original = target.log;
		const store = new LogStore();
		const unhook = hookConsole(target, store, { methods: ['log'] });

		target.log('hello', 1);
		unhook();
		target.log('after');

		expect(calls).toEqual([['hello', 1], ['after']]);
		expect(store.size).toBe(1);
		expect(target.log).toBe(original);
	});

	it('does not break the page when recording fails', () => {
		const { target } = fakeConsole();
		const store = new LogStore();

		vi.spyOn(store, 'append').mockImplementation(() => {
			throw new Error('store failure');
		});
		hookConsole(target, store, { methods: ['log'] });

		expect(() => target.log('still fine')).not.toThrow();
	});

	it('leaves a later wrapper in place and stops recording', () => {
		const { target } = fakeConsole();
		const store = new LogStore();
		const unhook = hookConsole(target, store, { methods: ['log'] });
		const ours = target.log;
		const later = (...args: unknown[]) => ours(...args);

		target.log = later;
		unhook();
		target.log('ignored');

		expect(target.log).toBe(later);
		expect(store.size).toBe(0);
	});

	it('can skip the original method', () => {
		const { target, calls } = fakeConsole();

		hookConsole(target, new LogStore(), { methods: ['warn'], passthrough: false });
		target.warn('quiet');

		expect(calls).toEqual([]);
	});
});

File: packages/flutter/test/sources/console_test.dart
import 'package:flutter/foundation.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:lognal/src/core/store.dart';
import 'package:lognal/src/core/types.dart';
import 'package:lognal/src/sources/console/capture.dart';
import 'package:lognal/src/sources/console/console.dart';
import 'package:lognal/src/sources/console/css_color.dart';
import 'package:lognal/src/sources/console/format.dart';
import 'package:lognal/src/sources/console/hook.dart';

ValueNode capture(Object? value) => captureValue(value);

String textOf(List<LogPart> parts) {
  return parts
      .map(
        (LogPart part) => part is TextPart ? part.text : '<${(part as ValuePart).value.kind.name}>',
      )
      .join();
}

LogEntry lastEntry(LogStore store) => store.at(store.size - 1)!;

class Account {
  Account(this.id);

  final int id;

  Map<String, Object?> toJson() => <String, Object?>{'id': id};
}

class Plain {
  const Plain();
}

class Described {
  const Described();

  @override
  String toString() => 'Described(all of it)';
}

void main() {
  group('captureValue', () {
    test('captures primitives the way they print', () {
      expect(captureValue(-0.0).kind, ValueKind.number);
      expect(captureValue(-0.0).value, '-0.0');
      expect(captureValue(BigInt.from(10)).value, '10');
      expect(captureValue(#token).kind, ValueKind.symbol);
      expect(captureValue(null).kind, ValueKind.nullValue);
      expect(captureValue(double.nan).value, 'NaN');
    });

    test('takes the value at call time', () {
      final Map<String, Object?> user = <String, Object?>{'name': 'before'};
      final ValueNode node = captureValue(user);

      user['name'] = 'after';

      expect(node.children?.first.value.value, 'before');
    });

    test('marks circular references', () {
      final Map<String, Object?> value = <String, Object?>{};

      value['self'] = value;

      expect(captureValue(value).children?.first.value.kind, ValueKind.circular);
    });

    test('respects the depth, property, string and node limits', () {
      const Map<String, Object?> deep = <String, Object?>{
        'a': <String, Object?>{
          'b': <String, Object?>{
            'c': <String, Object?>{'d': 1},
          },
        },
      };
      final Map<String, Object?> wide = <String, Object?>{
        for (int index = 0; index < 10; index++) 'key$index': index,
      };
      final ValueNode depthNode = captureValue(deep, const CaptureOptions(maxDepth: 2));
      final ValueNode? inner = depthNode.children?.first.value.children?.first.value;

      expect(inner?.kind, ValueKind.map);
      expect(inner?.children, isNull);
      expect(captureValue(wide, const CaptureOptions(maxProperties: 3)).omitted, 7);

      final ValueNode long = captureValue('x' * 20, const CaptureOptions(maxStringLength: 5));

      expect(long.value, 'xxxxx');
      expect(long.truncated, 15);
      expect(
        captureValue(
          List<Map<String, int>>.generate(50, (int _) => <String, int>{'x': 1}),
          const CaptureOptions(maxNodes: 10),
        ).children?.length,
        lessThan(10),
      );
    });

    test('recognizes built-in types', () {
      expect(captureValue(DateTime.utc(1970)).kind, ValueKind.date);
      expect(captureValue(DateTime.utc(1970)).value, '1970-01-01T00:00:00.000Z');
      expect(captureValue(RegExp('a+b')).value, '/a+b/');
      expect(captureValue(<String, int>{'a': 1}).kind, ValueKind.map);
      expect(captureValue(<String, int>{'a': 1}).size, 1);
      expect(captureValue(<int>{1, 2}).kind, ValueKind.set);
      expect(captureValue(<int>[1, 2, 3]).size, 3);
      expect(captureValue(Uint8List.fromList(<int>[1, 2, 3])).kind, ValueKind.list);
      expect(captureValue(int).kind, ValueKind.classValue);
      expect(captureValue(main).kind, ValueKind.function);
      expect(captureValue(Future<int>.value(1)).kind, ValueKind.future);
    });

    test('keeps the type name of an instance and opens it through toJson', () {
      final ValueNode node = captureValue(Account(7));

      expect(node.className, 'Account');
      expect(node.children?.first.key, 'id');
      expect(node.children?.first.value.value, '7');
      expect(captureValue(Account(7), const CaptureOptions(expandToJson: false)).children, isNull);
    });

    test('shows an object that cannot be opened by what it says about itself', () {
      expect(captureValue(const Plain()).value, isNull);
      expect(captureValue(const Described()).value, 'Described(all of it)');
    });

    test('captures errors with their message and their stack', () {
      late ValueNode node;

      try {
        throw StateError('broken');
      } on StateError catch (error, stack) {
        node = captureValue(error).copyWith(stack: stack.toString());
      }

      expect(node.kind, ValueKind.error);
      expect(node.className, 'StateError');
      expect(node.value, 'broken');
      expect(node.stack, isNotEmpty);
      expect(captureValue(const FormatException('bad')).value, 'bad');
    });

    test('survives a toJson that throws', () {
      expect(() => captureValue(_Hostile()), returnsNormally);
      expect(captureValue(_Hostile()).className, '_Hostile');
    });
  });

  group('format specifiers', () {
    test('applies %s, %d, %i and %f the way the Console Standard describes', () {
      final FormattedMessage formatted = applyFormat(<Object?>[
        '%s|%d|%i|%f|%%',
        'text',
        '42.9',
        -1.5,
        '3.25',
        'extra',
      ], capture);

      expect(textOf(formatted.parts), 'text|42|-1|3.25|%');
      expect(formatted.rest, <Object?>['extra']);
    });

    test('inserts values for %o and %O and leaves specifiers without arguments as written', () {
      final FormattedMessage formatted = applyFormat(<Object?>[
        'a %o b %O c %s',
        <String, int>{'x': 1},
        <int>[1],
      ], capture);

      expect(textOf(formatted.parts), 'a <map> b <list> c %s');
    });

    test('styles the text after %c with the allowed properties only', () {
      final FormattedMessage formatted = applyFormat(<Object?>[
        '%cred%c plain',
        'color: red; background: url(https://example.com/x.png) #000; font-weight: bold',
        '',
      ], capture);
      final TextPart first = formatted.parts.first as TextPart;

      expect(first.text, 'red');
      expect(first.style?.color, const RgbTextColor(0xffff0000));
      expect(first.style?.background, const RgbTextColor(0xff000000));
      expect(first.style?.bold, isTrue);
      expect((formatted.parts[1] as TextPart).style, isNull);
    });

    test('rejects colors that could load a resource', () {
      expect(parseCssColor('rgb(1 2 3 / 50%)'), isNotNull);
      expect(parseCssColor('url(x)'), isNull);
      expect(parseConsoleCss('color: url(x)'), isNull);
    });

    test('joins the remaining arguments with spaces', () {
      expect(
        textOf(formatArguments(<Object?>['count', 3, 'items'], capture)),
        'count <number> items',
      );
      expect(textOf(formatArguments(<Object?>['%d%%', 50, 'done'], capture)), '50% done');
      expect(textOf(formatArguments(<Object?>['only %s'], capture)), 'only %s');
    });
  });

  group('LognalConsole', () {
    test('records levels and group membership', () {
      final LogStore store = LogStore();
      final LognalConsole log = LognalConsole(store);

      log.group('outer');
      log.warn('inside');
      log.groupEnd();
      log.error('outside');

      final List<LogEntry> entries = store.toList();

      expect(entries[0].kind, LogKind.group);
      expect(entries[0].collapsed, isFalse);
      expect(entries[1].level, LogLevel.warn);
      expect(entries[1].groups, <int>[entries[0].id]);
      expect(entries[2].level, LogLevel.error);
      expect(entries[2].groups, isEmpty);
    });

    test('counts, resets counts and warns about a missing counter', () {
      final LogStore store = LogStore(
        options: const LogStoreOptions(mergeRepeats: RepeatMode.keep),
      );
      final LognalConsole log = LognalConsole(store);

      log.count('clicks');
      log.count('clicks');
      log.countReset('clicks');
      log.count('clicks');
      log.countReset('other');

      expect(store.toList().map((LogEntry entry) => textOf(entry.parts)).toList(), <String>[
        'clicks: 1',
        'clicks: 2',
        'clicks: 1',
        "Count for 'other' does not exist",
      ]);
    });

    test('measures timers and warns about a missing one', () {
      final LogStore store = LogStore();
      final LognalConsole log = LognalConsole(store);

      log.time('load');
      log.timeEnd('load');
      log.timeEnd('load');

      final List<String> texts = store
          .toList()
          .map((LogEntry entry) => textOf(entry.parts))
          .toList();

      expect(texts.first, startsWith('load: '));
      expect(texts.first, endsWith(' ms'));
      expect(texts[1], "Timer 'load' does not exist");
    });

    test('writes assertion failures the way the standard describes', () {
      final LogStore store = LogStore(
        options: const LogStoreOptions(mergeRepeats: RepeatMode.keep),
      );
      final LognalConsole log = LognalConsole(store);

      log.assertCondition(true, <Object?>['not shown']);
      log.assertCondition(false);
      log.assertCondition(false, <Object?>['value is %d', 3]);
      log.assertCondition(false, <Object?>[
        <String, int>{'code': 1},
      ]);

      expect(store.toList().map((LogEntry entry) => textOf(entry.parts)).toList(), <String>[
        'Assertion failed',
        'Assertion failed: value is 3',
        'Assertion failed <map>',
      ]);
    });

    test('draws a table that lines up wide characters', () {
      final LogStore store = LogStore();

      LognalConsole(store).table(<Map<String, Object?>>[
        <String, Object?>{'name': '김철수', 'age': 30},
        <String, Object?>{'name': 'Ann'},
      ]);

      final List<String> lines = textOf(lastEntry(store).parts).split('\n');
      final Set<int> widths = lines.map((String line) {
        return line.runes.fold<int>(
          0,
          (int sum, int rune) => sum + (rune >= 0xac00 && rune <= 0xd7a3 ? 2 : 1),
        );
      }).toSet();

      expect(lines[1], contains('(index)'));
      expect(widths, hasLength(1));
    });

    test('clears the store and notes it', () {
      final LogStore store = LogStore();
      final LognalConsole log = LognalConsole(store);

      log.log('before');
      log.clear();

      expect(store.size, 1);
      expect(lastEntry(store).kind, LogKind.system);
    });

    test('adds the stack of trace', () {
      final LogStore store = LogStore();

      LognalConsole(store).trace('here');

      final String text = textOf(lastEntry(store).parts);

      expect(text, startsWith('here\n'));
      expect(text, contains('console_test.dart'));
    });
  });

  group('hookDebugPrint', () {
    test('records calls, still runs the original and restores it', () {
      final List<String?> printed = <String?>[];
      final DebugPrintCallback original = debugPrint;
      final LogStore store = LogStore();

      debugPrint = (String? message, {int? wrapWidth}) => printed.add(message);

      final DebugPrintCallback ours = debugPrint;
      final void Function() unhook = hookDebugPrint(store);

      debugPrint('hello');
      unhook();
      debugPrint('after');

      expect(printed, <String?>['hello', 'after']);
      expect(store.size, 1);
      expect(debugPrint, same(ours));

      debugPrint = original;
    });

    test('leaves a later wrapper in place and stops recording', () {
      final DebugPrintCallback original = debugPrint;
      final LogStore store = LogStore();
      final void Function() unhook = hookDebugPrint(
        store,
        options: const HookOptions(passthrough: false),
      );
      final DebugPrintCallback ours = debugPrint;

      void later(String? message, {int? wrapWidth}) => ours(message, wrapWidth: wrapWidth);

      debugPrint = later;
      unhook();
      debugPrint('ignored');

      expect(debugPrint, same(later));
      expect(store.size, 0);

      debugPrint = original;
    });

    test('runs code in a zone whose print reaches the store', () {
      final LogStore store = LogStore();

      runZonedWithLognal(store, () {
        // ignore: avoid_print
        print('from the zone');
      }, options: const HookOptions(passthrough: false));

      expect(store.size, 1);
      expect(textOf(store.at(0)!.parts), 'from the zone');
    });
  });
}

class _Hostile {
  Map<String, Object?> toJson() => throw StateError('no');
}
